Ansible module introspect

alikins opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/module_utils

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (cowsay_py 9fe6a5a3f5) last updated 2017/01/25 16:01:30 (GMT -400)
  config file = /home/adrian/ansible/ansible.cfg
  configured module search path = [u'/home/adrian/ansible/my-modules', u'/home/adrian/ansible/lib/modules/']


```

##### SUMMARY
Adds support for module introspection. If invoked with the right options, the json returned from a python ansible module  will include an assortment of additional info about the module, it's args and values, and the environment and runtime it ran in. 

The new info goes into an 'introspection' key in the retrurned json (aside facts and invocation).

A couple of use cases.

1) running a module and introspecting it's arg spec (returned as a data structure here). In combo with 'introspect_only' this can help tools that want to dynamically inspect a modules args

 The info can also be returned in additional to normal module return data. So a tool like Tower that invoked modules and records the results, the arg spec could be used to generate a 'run again with different args' page. Or possibly the CLI could augment 'retry' files.

2) Module debugging and troubleshooting

  The introspection data includes a lot of info that usually requires ANSIBLE_KEEP_REMOTE_FILES, module 'explode', editing the module source, and running again. 

 For troubleshooting something like 'is this using the right python module for libfoobarblip?', flipping on introspection and seeing python sys.path and 'site' options is much quicker. 

 Other info like uid, euid, pid, module file name, the default args used, etc will be useful for debugging/troubleshooting.

3) event logging and auditting

  This adds a lot of context info about a module/task and does it with structured data. Using that info to populate something like an ELK stack would provide a lot of historical info for later review.

4) playbook troubleshooting

  It can be hard to figure out what default args end up being used when a module is invoked. The introspection info provides a lot of extra context to help.

An example invocation:


```
[fedora-25:ansible (ansible_module_introspect %)]$ ANSIBLE_MODULE_INTROSPECT=1 bin/ansible -i hosts grimlock -m copy -a 'src=/tmp/test_file dest=/tmp/setup.py' -C -D
```

``` json
grimlock | SUCCESS => {
    "changed": false, 
    "checksum": "4fbebcd2162b75b9e8f490397c148d4ae635a32a", 
    "dest": "/tmp/setup.py", 
    "gid": 1002, 
    "group": "adrian", 
    "introspect": {
        "module": {
            "aliases": {
                "attr": "attributes", 
                "dest": "path", 
                "name": "path"
            }, 
            "ansible_version": "2.3.0", 
            "argument_spec": {
                "attributes": {
                    "aliases": [
                        "attr"
                    ]
                }, 
                "backup": {}, 
                "content": {
                    "no_log": true
                }, 
                "delimiter": {}, 
                "diff_peek": {
                    "default": null
                }, 
                "directory_mode": {}, 
                "follow": {
                    "default": false, 
                    "type": "bool"
                }, 
                "force": {
                    "default": false, 
                    "required": false, 
                    "type": "bool"
                }, 
                "group": {}, 
                "mode": {
                    "type": "raw"
                }, 
                "original_basename": {
                    "required": false
                }, 
                "owner": {}, 
                "path": {
                    "aliases": [
                        "dest", 
                        "name"
                    ], 
                    "required": true, 
                    "type": "path"
                }, 
                "recurse": {
                    "default": false, 
                    "type": "bool"
                }, 
                "regexp": {}, 
                "remote_src": {}, 
                "selevel": {}, 
                "serole": {}, 
                "setype": {}, 
                "seuser": {}, 
                "src": {
                    "default": null, 
                    "required": false, 
                    "type": "path"
                }, 
                "state": {
                    "choices": [
                        "file", 
                        "directory", 
                        "link", 
                        "hard", 
                        "touch", 
                        "absent"
                    ], 
                    "default": null
                }, 
                "unsafe_writes": {
                    "type": "bool"
                }, 
                "validate": {
                    "default": null, 
                    "required": false
                }
            }, 
            "check_mode": true, 
            "cleanup_files": [], 
            "debug": false, 
            "diff": true, 
            "legal_inputs": [
                "_ansible_check_mode", 
                "_ansible_no_log", 
                "_ansible_debug", 
                "_ansible_diff", 
                "_ansible_verbosity", 
                "_ansible_selinux_special_fs", 
                "_ansible_module_name", 
                "_ansible_version", 
                "_ansible_syslog_facility", 
                "_ansible_module_introspect", 
                "_ansible_module_introspect_only", 
                "directory_mode", 
                "force", 
                "remote_src", 
                "path", 
                "dest", 
                "name", 
                "owner", 
                "follow", 
                "group", 
                "unsafe_writes", 
                "state", 
                "content", 
                "serole", 
                "diff_peek", 
                "setype", 
                "selevel", 
                "original_basename", 
                "regexp", 
                "validate", 
                "src", 
                "seuser", 
                "recurse", 
                "delimiter", 
                "mode", 
                "attributes", 
                "attr", 
                "backup"
            ], 
            "metadata": {
                "documentation": "\n---\nmodule: file\nversion_added: \"historical\"\nshort_description: Sets attributes of files\nextends_documentation_fragment: files\ndescription:\n     - Sets attributes of files, symlinks, and directories, or removes\n       files/symlinks/directories. Many other modules support the same options as\n       the M(file) module - including M(copy), M(template), and M(assemble).\nnotes:\n    - See also M(copy), M(template), M(assemble)\nrequirements: [ ]\nauthor:\n    - \"Ansible Core Team\"\n    - \"Michael DeHaan\"\noptions:\n  path:\n    description:\n      - 'path to the file being managed.  Aliases: I(dest), I(name)'\n    required: true\n    default: []\n    aliases: ['dest', 'name']\n  state:\n    description:\n      - If C(directory), all immediate subdirectories will be created if they\n        do not exist, since 1.7 they will be created with the supplied permissions.\n        If C(file), the file will NOT be created if it does not exist, see the M(copy)\n        or M(template) module if you want that behavior.  If C(link), the symbolic\n        link will be created or changed. Use C(hard) for hardlinks. If C(absent),\n        directories will be recursively deleted, and files or symlinks will be unlinked.\n        Note that M(file) will not fail if the C(path) does not exist as the state did not change.\n        If C(touch) (new in 1.4), an empty file will be created if the C(path) does not\n        exist, while an existing file or directory will receive updated file access and\n        modification times (similar to the way `touch` works from the command line).\n    required: false\n    default: file\n    choices: [ file, link, directory, hard, touch, absent ]\n  src:\n    required: false\n    default: null\n    description:\n      - path of the file to link to (applies only to C(state=link)). Will accept absolute,\n        relative and nonexisting paths. Relative paths are not expanded.\n  recurse:\n    required: false\n    default: \"no\"\n    choices: [ \"yes\", \"no\" ]\n    version_added: \"1.1\"\n    description:\n      - recursively set the specified file attributes (applies only to state=directory)\n  force:\n    required: false\n    default: \"no\"\n    choices: [ \"yes\", \"no\" ]\n    description:\n      - 'force the creation of the symlinks in two cases: the source file does\n        not exist (but will appear later); the destination exists and is a file (so, we need to unlink the\n        \"path\" file and create symlink to the \"src\" file in place of it).'\n  follow:\n    required: false\n    default: \"no\"\n    choices: [ \"yes\", \"no\" ]\n    version_added: \"1.8\"\n    description:\n      - 'This flag indicates that filesystem links, if they exist, should be followed.'\n", 
                "examples": "\n# change file ownership, group and mode. When specifying mode using octal numbers, first digit should always be 0.\n- file:\n    path: /etc/foo.conf\n    owner: foo\n    group: foo\n    mode: 0644\n- file:\n    src: /file/to/link/to\n    dest: /path/to/symlink\n    owner: foo\n    group: foo\n    state: link\n- file:\n    src: '/tmp/{{ item.src }}'\n    dest: '{{ item.dest }}'\n    state: link\n  with_items:\n    - { src: 'x', dest: 'y' }\n    - { src: 'z', dest: 'k' }\n\n# touch a file, using symbolic modes to set the permissions (equivalent to 0644)\n- file:\n    path: /etc/foo.conf\n    state: touch\n    mode: \"u=rw,g=r,o=r\"\n\n# touch the same file, but add/remove some permissions\n- file:\n    path: /etc/foo.conf\n    state: touch\n    mode: \"u+rw,g-wx,o-rwx\"\n\n# create a directory if it doesn't exist\n- file:\n    path: /etc/some_directory\n    state: directory\n    mode: 0755\n", 
                "metadata": {
                    "status": [
                        "stableinterface"
                    ], 
                    "supported_by": "core", 
                    "version": "1.0"
                }, 
                "return": null
            }, 
            "module_path": "/tmp/ansible_5vOSC5/ansible_modlib.zip/ansible/module_utils", 
            "name": "file", 
            "no_log": false, 
            "no_log_values": [], 
            "params": {
                "attributes": null, 
                "backup": null, 
                "content": null, 
                "delimiter": null, 
                "dest": "/tmp/setup.py", 
                "diff_peek": null, 
                "directory_mode": null, 
                "follow": false, 
                "force": false, 
                "group": null, 
                "mode": null, 
                "original_basename": "test_file", 
                "owner": null, 
                "path": "/tmp/setup.py", 
                "recurse": false, 
                "regexp": null, 
                "remote_src": null, 
                "selevel": null, 
                "serole": null, 
                "setype": null, 
                "seuser": null, 
                "src": "test_file", 
                "state": null, 
                "unsafe_writes": null, 
                "validate": null
            }, 
            "run_command_environ_update": {}, 
            "supported_params": [
                "directory_mode", 
                "force", 
                "remote_src", 
                "path", 
                "dest", 
                "name", 
                "owner", 
                "follow", 
                "group", 
                "unsafe_writes", 
                "state", 
                "content", 
                "serole", 
                "diff_peek", 
                "setype", 
                "selevel", 
                "original_basename", 
                "regexp", 
                "validate", 
                "src", 
                "seuser", 
                "recurse", 
                "delimiter", 
                "mode", 
                "attributes", 
                "attr", 
                "backup"
            ], 
            "supports_check_mode": true, 
            "used_defaults": {
                "attributes": null, 
                "backup": null, 
                "content": null, 
                "delimiter": null, 
                "diff_peek": null, 
                "directory_mode": null, 
                "follow": false, 
                "force": false, 
                "group": null, 
                "mode": null, 
                "owner": null, 
                "recurse": false, 
                "regexp": null, 
                "remote_src": null, 
                "selevel": null, 
                "serole": null, 
                "setype": null, 
                "seuser": null, 
                "state": null, 
                "unsafe_writes": null, 
                "validate": null
            }, 
            "used_fallbacks": {}, 
            "verbosity": 0
        }, 
        "module_invocation": {
            "environment": {
                "CVS_RSH": "ssh", 
                "EDITOR": "vim -f",
                "GIT_PROMPT": "/usr/share/git-core/contrib/completion/git-prompt.sh", 
                "GIT_PS1_SHOWCOLORHINTS": "1", 
                "GIT_PS1_SHOWUNTRACKEDFILES": "1", 
                "GIT_PS1_SHOWUPSTREAM": "verbose", 
                "GOPATH": "/home/adrian/gopath", 
                "GUESTFISH_INIT": "\\e[1;34m", 
                "GUESTFISH_OUTPUT": "\\e[0m", 
                "GUESTFISH_PS1": "\\[\\e[1;32m\\]><fs>\\[\\e[0;31m\\] ", 
                "GUESTFISH_RESTORE": "\\e[0m", 
                "HISTFILESIZE": "500000", 
                "HISTSIZE": "300000", 
                "HISTTIMEFORMAT": "%F %T ", 
                "HOME": "/home/adrian", 
                "HOSTNAME": "grimlock", 
                "ID": "rhel", 
                "JAVA_HOME": "/usr/lib/jvm/java", 
                "KDEDIRS": "/usr", 
                "LANG": "en_US.utf8", 
                "LC_MEASUREMENT": "en_US.utf8", 
                "LC_MONETARY": "en_US.utf8", 
                "LC_NUMERIC": "en_US.utf8", 
                "LC_PAPER": "en_US.utf8", 
                "LC_TIME": "en_US.utf8", 
                "LESSOPEN": "||/usr/bin/lesspipe.sh %s", 
                "LOGNAME": "adrian", 
                "M2_REPO": "/home/adrian/.m2/repository", 
                "MAIL": "/var/mail/adrian", 
                "PATH": "/home/adrian/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/bin:/home/adrian/.cabal/bin:/home/adrian/gopath/bin:/home/adrian/.local/bin", 
                "PKG_CONFIG_PATH": "/usr/local/lib/pkgconfig/", 
                "PWD": "/home/adrian", 
                "PYTHONPATH": "/tmp/ansible_5vOSC5/ansible_modlib.zip", 
                "PYTHONSTARTUP": "/home/adrian/.pystartup", 
                "QTDIR": "/usr/lib64/qt-3.3", 
                "QTINC": "/usr/lib64/qt-3.3/include", 
                "QTLIB": "/usr/lib64/qt-3.3/lib", 
                "QT_GRAPHICSSYSTEM_CHECKED": "1", 
                "QT_PLUGIN_PATH": "/usr/lib64/kde4/plugins:/usr/lib/kde4/plugins", 
                "SELINUX_LEVEL_REQUESTED": "", 
                "SELINUX_ROLE_REQUESTED": "", 
                "SELINUX_USE_CURRENT_RANGE": "", 
                "SHELL": "/bin/bash", 
                "SHLVL": "2", 
                "SSH_ASKPASS": "/usr/libexec/openssh/gnome-ssh-askpass", 
                "SSH_AUTH_SOCK": "/tmp/ssh-EWP2ppAm1O/agent.3298", 
                "SSH_CLIENT": "192.168.1.35 40784 22", 
                "SSH_CONNECTION": "192.168.1.35 40784 192.168.1.36 22", 
                "USER": "adrian", 
                "USER_LS_COLORS": "", 
                "VAGRANT_DEFAULT_PROVIDER": "libvirt", 
                "VERSION_ID": "7.2", 
                "VIRTUALENVWRAPPER_HOOK_DIR": "/home/adrian/.virtualenvs", 
                "VIRTUALENVWRAPPER_PROJECT_FILENAME": ".project", 
                "VIRTUALENVWRAPPER_SCRIPT": "/home/adrian/bin/virtualenvwrapper.sh", 
                "WORKON_HOME": "/home/adrian/.virtualenvs", 
                "XDG_RUNTIME_DIR": "/run/user/1000", 
                "XDG_SESSION_ID": "131323", 
                "XMODIFIERS": "@im=ibus", 
                "_": "/usr/bin/python"
            }, 
            "locale": [
                "en_US", 
                "UTF-8"
            ], 
            "process": {
                "cwd": "/home/adrian", 
                "euid": 1000, 
                "gid": 1002, 
                "pid": 3688, 
                "ppid": 3687, 
                "uid": 1000
            }, 
            "python": {
                "sys_argv": [
                    "/tmp/ansible_5vOSC5/ansible_module_file.py"
                ], 
                "sys_exec_prefix": "/usr", 
                "sys_meta_path": [], 
                "sys_path": [
                    "/tmp/ansible_5vOSC5", 
                    "/tmp/ansible_5vOSC5/ansible_modlib.zip", 
                    "/home/adrian/.local/lib/python2.7/site-packages/_pdbpp_path_hack", 
                    "/home/adrian/.local/lib/python2.7/site-packages/Gelatin-DEVELOPMENT-py2.7.egg", 
                    "/home/adrian/.local/lib/python2.7/site-packages/keyring-2.0dev-py2.7.egg", 
                    "/home/adrian/.local/lib/python2.7/site-packages/ansible_lint-2.1.2-py2.7.egg", 
                    "/home/adrian/.local/lib/python2.7/site-packages/github_comments-1.3-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/yolk-0.4.3-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/rho-0.0.21-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/tbgrep-0.2.3-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/translate_toolkit-1.13.0_alpha1-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/centpkg-0.4.4-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/rainbow_logging_handler-2.2.2-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/bkr.systemscan-1.6-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/github_comments-1.3-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/noselog-0.0.5-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/tito-0.6.2-py2.7.egg", 
                    "/usr/lib/python2.7/site-packages/maproxy-0.0.12-py2.7.egg", 
                    "/home/adrian/.local/lib/python2.7/site-packages/subscription_manager-1.17.3-py2.7.egg", 
                    "/tmp/ansible_5vOSC5/ansible_modlib.zip", 
                    "/usr/lib64/python27.zip", 
                    "/usr/lib64/python2.7", 
                    "/usr/lib64/python2.7/plat-linux2", 
                    "/usr/lib64/python2.7/lib-tk", 
                    "/usr/lib64/python2.7/lib-old", 
                    "/usr/lib64/python2.7/lib-dynload", 
                    "/home/adrian/.local/lib/python2.7/site-packages", 
                    "/usr/lib/python2.7/site-packages", 
                    "/usr/lib64/python2.7/site-packages", 
                    "/usr/lib64/python2.7/site-packages/gtk-2.0"
                ], 
                "sys_path_hooks": [
                    "<type 'zipimport.zipimporter'>"
                ], 
                "sys_prefix": "/usr"
            }, 
            "python_site_info": {
                "site_PREFIXES": [
                    "/usr", 
                    "/usr"
                ], 
                "site_USER_BASE": "/home/adrian/.local", 
                "site_USER_SITE": "/home/adrian/.local/lib/python2.7/site-packages", 
                "site_getsitepackages": [
                    "/usr/lib64/python2.7/site-packages", 
                    "/usr/lib/python2.7/site-packages", 
                    "/usr/lib/site-python"
                ]
            }
        }
    }, 
    "mode": "0664", 
    "owner": "adrian", 
    "path": "/tmp/setup.py", 
    "secontext": "unconfined_u:object_r:user_tmp_t:s0", 
    "size": 2077, 
    "state": "file", 
    "uid": 1000
}
```

alikins committed:
Add a 'introspect' to AnsibleModule

If _ansible_module_introspect module args is truthy, add
some module introspection data into the 'introspect' of the
returned json.

alikins committed:
Add a DEFAULT_MODULE_INTROSPECT

Allow ANSIBLE_MODULE_INTROSPECT to pass args to module
incovation telling it to return module introspection data.

alikins committed:
poke around modules __main__ instead of passing in metadata

alikins committed:
Add a 'introspect only mode'

alikins committed:
clean up diff noise

alikins committed:
setup user_defaults/fallbacks earlier

better error message for invalid args

alikins committed:
getstate->_module_introspect, split __main__ stuff to method

alikins committed:
better name for module info introspect method

This code is split into a method so that it can easily
be overridden or mocked, since the current implementation
relies on __main__ being a normal ansible module.

alikins committed:
move __main__ import to module scope

And import it as 'module_main'.
Eventually would like a way to specify a provider for
the module info (possibly as info passed as args to
AnsibleModule init()).

alikins committed:
notes/comments

alikins committed:
split up introspect stuff

Add info about module invocation (the current process) including
python setup/paths, pid/cwd/*id, and current env vars

alikins committed:
split "static" introspection stuff to its own module

alikins committed:
Add sys.modules info to introspection

alikins committed:
add python version info based on facts.py

alikins committed:
split python introspec to small chunks

Mainly so facts only needs to call one method from here

alikins committed:
update facts.get_python_facts to use 'introspection'

ryansb commented:
This works for me on simple modules (ping, copy, etc), but using a module with external libs fails. This is one from testing with `docker_login`.

```
Play:
    - docker_login:
        registry_url: [secrets]
        username: ryansb
        password: [secrets
Traceback:
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ryansb/ansible_XyBRCL/ansible_module_docker_login.py", line 333, in <module>
    main()
  File "/tmp/ryansb/ansible_XyBRCL/ansible_module_docker_login.py", line 327, in main
    client.module.exit_json(**results)
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 1985, in exit_json
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in remove_values
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in <genexpr>
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in remove_values
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in <genexpr>
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in remove_values
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in <genexpr>
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in remove_values
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in <genexpr>
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in remove_values
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 414, in <genexpr>
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 412, in remove_values
  File "/tmp/ryansb/ansible_XyBRCL/ansible_modlib.zip/ansible/module_utils/basic.py", line 425, in remove_values
TypeError: Value of unknown type: <class 'requests.packages.urllib3.packages.six._SixMetaPathImporter'>, <requests.packages.urllib3.packages
.six._SixMetaPathImporter object at 0x7f33ef984e90>
```

alikins committed:
merge cleanup

alikins commented:
> TypeError: Value of unknown type: <class 'requests.packages.urllib3.packages.six._SixMetaPathImporter'>, <requests.packages.urllib3.packages
> .six._SixMetaPathImporter object at 0x7f33ef984e90>

Guessing the _SixMetaPathImporter is not serializable, looks like it needs the same handling as sys.path_hooks.

alikins committed:
repr() sys.meta_path entries before serializing

alikins committed:
move implementation to introspection.py

And guard use of introspection.py with import sentinels so
we can handle the introspection module being missing.

alikins committed:
Include the 'introspection' module iff we are using it

Check the module args for '_ansible_module_introspect' and add
the introspection module to the ansiballz if enabled.

Would be better if we could prevent it from being added to the module
list for the disabled case, instead of having to remove it. tbd

alikins commented:
meh
